zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs

```diff
@@ -131,7 +135,7 @@ impl<F: Field> ExecutionGadget<F> for CallDataCopyGadget<F> {
                 src_addr,
                 src_addr_end,
                 memory_address.offset(),
-                memory_address.length(),
+                bytes_length_word.expr(),
```

Comment:
The copy circuit is going to need the real copy length (like currently). Otherwise, it would not know what the actual copy range is, and it will copy too many bytes. The circuit does not really need the total padded length, it can check it internally with the word index.

I think the calls to `copy_table_lookup` do not need to change.

Reply:
yes, I understand your intention. originally I want to remain the `copy_table_lookup` call in opcode/evm circuit, but `copy_table_lookup` call essentially lookup the first row of copy event. and `bytes_length_word` is corresponding to `bytes_left` column of copy table. copy table use the all the bytes length at the first row, and then decrease in later consecutive rows   if use original len in opcode gadget, `copy_table_lookup` call will fail.

Reply:
Yes, the circuit does need to manage the total height of the copy event somehow, as well as the rows for both the front and the back padding, for both the source and the destination. Important from the execution circuit is the copy length which is a parameter of EVM instructions.

The circuit `bytes_left` does not need to count the total length, it can count the real copy length (then it does not change during padding).

Reply:
if do like this, need to change `bytes_left` logic in copy circuit, can not get it directly from `step_idx`. the advantage is no change in evm circuit. let's think the detail.

Reply:
fixed in [2cc3672](https://github.com/scroll-tech/zkevm-circuits/pull/466/commits/2cc3672ce6084071b7609a44369131cce223cc9f)